File: src/velox/api/layer0/live/DemoExternalRealtimeProvider.java
package velox.api.layer0.live;

import java.util.HashMap;

import velox.api.layer1.Layer1ApiAdminListener;
import velox.api.layer1.data.InstrumentInfo;
import velox.api.layer1.data.LoginData;
import velox.api.layer1.data.LoginFailedReason;
import velox.api.layer1.data.OrderSendParameters;
import velox.api.layer1.data.OrderUpdateParameters;
import velox.api.layer1.data.TradeInfo;
import velox.api.layer1.data.UserPasswordDemoLoginData;

/**
 * <p>
 * This a demo provider that generates data instead of actually receiving it.
 * </p>
 */
public class DemoExternalRealtimeProvider extends ExternalLiveBaseProvider {

    protected class Instrument {
        /** Number of depth levels that will be generated on each side */
        private static final int DEPTH_LEVELS_COUNT = 10;

        protected final String alias;
        protected final double pips;

        private int basePrice;

        public Instrument(String alias, double pips) {
            this.alias = alias;
            this.pips = pips;

            // Pick random price that will be used to generate the data
            // This is an integer representation of a price (before multiplying
            // by pips)
            this.basePrice = (int) (Math.random() * 10000 + 1000);
        }

        public void generateData() {

            // Determining best bid/ask
            int bestBid = getBestBid();
            int bestAsk = getBestAsk();

            // Populating 10 levels to each side of best bid/best ask with
            // random data
            for (int i = 0; i < DEPTH_LEVELS_COUNT; ++i) {
                final int levelsOffset = i;
                dataListeners.forEach(l -> l.onDepth(alias, true, bestBid - levelsOffset, getRandomSize()));
                dataListeners.forEach(l -> l.onDepth(alias, false, bestAsk + levelsOffset, getRandomSize()));
            }

            // Currently Bookmap does not visualize OTC trades, so you will
            // mostly want isOtc=false
            final boolean isOtc = false;
            // Trade on best bid, ask agressor
            dataListeners.forEach(l -> l.onTrade(alias, bestBid, 1, new TradeInfo(isOtc, false)));
            // Trade on best ask, bid agressor
            dataListeners.forEach(l -> l.onTrade(alias, bestAsk, 1, new TradeInfo(isOtc, true)));

            // With 10% chance change BBO
            if (Math.random() < 0.1) {
                // 50% chance to move up, 50% to move down
                if (Math.random() > 0.5) {
                    // Moving up - erasing best ask, erasing last reported bid
                    // level (emulating exchange only reporting few levels)
                    ++basePrice;
                    dataListeners.forEach(l -> l.onDepth(alias, false, bestAsk, 0));
                    dataListeners.forEach(l -> l.onDepth(alias, true, bestBid - (DEPTH_LEVELS_COUNT - 1), 0));
                    // Could also populate new best bid and add last best ask,
                    // but this can be omitted - those will be populated during
                    // next simulation step
                } else {
                    // Moving down - erasing best bid, erasing last reported ask
                    // level (emulating exchange only reporting few levels)
                    --basePrice;
                    dataListeners.forEach(l -> l.onDepth(alias, true, bestBid, 0));
                    dataListeners.forEach(l -> l.onDepth(alias, false, bestAsk + (DEPTH_LEVELS_COUNT - 1), 0));
                    // Could also populate new best ask and add last best bid,
                    // but this can be omitted - those will be populated during
                    // next simulation step
                }
            }
        }

        public int getBestAsk() {
            return basePrice;
        }

        public int getBestBid() {
            return getBestAsk() - 1;
        }

        private int getRandomSize() {
            return (int) (1 + Math.random() * 10);
        }

    }

    protected HashMap<String, Instrument> instruments = new HashMap<>();

    // This thread will perform data generation.
    private Thread connectionThread = null;

    /**
     * <p>
     * Generates alias from symbol, exchange and type of the instrument. Alias
     * is a unique identifier for the instrument, but it's also used in many
     * places in UI, so it should also be easily readable.
     * </p>
     * <p>
     * Note, that you don't have to use all 3 fields. You can just ignore some
     * of those, for example use symbol only.
     * </p>
     */
    private static String createAlias(String symbol, String exchange, String type) {
        return symbol + "/" + exchange + "/" + type;
    }

    @Override
    public void subscribe(String symbol, String exchange, String type) {
        String alias = createAlias(symbol, exchange, type);
        // Since instruments also will be accessed from the data generation
        // thread, synchronization is required
        //
        // No need to worry about calling listener from synchronized block,
        // since those will be processed asynchronously
        synchronized (instruments) {
            if (instruments.containsKey(alias)) {
                instrumentListeners.forEach(l -> l.onInstrumentAlreadySubscribed(symbol, exchange, type));
            } else {
                // We are performing subscription synchronously for simplicity,
                // but if subscription process takes long it's better to do it
                // asynchronously (e.g use Executor)

                // Randomly determining pips. In reality it will be received
                // from external source
                double pips = Math.random() > 0.5 ? 0.5 : 0.25;

                final Instrument newInstrument = new Instrument(alias, pips);
                instruments.put(alias, newInstrument);

                final InstrumentInfo instrumentInfo = new InstrumentInfo(
                        symbol, exchange, type, newInstrument.pips, 1, "", false);

                instrumentListeners.forEach(l -> l.onInstrumentAdded(alias, instrumentInfo));
            }
        }
    }

    @Override
    public void unsubscribe(String alias) {
        synchronized (instruments) {
            if (instruments.remove(alias) != null) {
                instrumentListeners.forEach(l -> l.onInstrumentRemoved(alias));
            }
        }
    }

    @Override
    public String formatPrice(String alias, double price) {
        // Use default Bookmap price formatting logic for simplicity.
        // Values returned by this method will be used on price axis and in few
        // other places.

        double pips;
        synchronized (instruments) {
            pips = instruments.get(alias).pips;
        }

        return formatPriceDefault(pips, price);
    }

    @Override
    public void sendOrder(OrderSendParameters orderSendParameters) {
        // This method will not be called because this adapter does not report
        // trading capabilities
        throw new RuntimeException("Not trading capable");
    }

    @Override
    public void updateOrder(OrderUpdateParameters orderUpdateParameters) {
        // This method will not be called because this adapter does not report
        // trading capabilities
        throw new RuntimeException("Not trading capable");
    }

    @Override
    public void login(LoginData loginData) {
        UserPasswordDemoLoginData userPasswordDemoLoginData = (UserPasswordDemoLoginData) loginData;

        // If connection process takes a while then it's better to do it in
        // separate thread
        connectionThread = new Thread(() -> handleLogin(userPasswordDemoLoginData));
        connectionThread.start();
    }

    private void handleLogin(UserPasswordDemoLoginData userPasswordDemoLoginData) {
        // With real connection provider would attempt establishing connection
        // here.
        boolean isValid = "pass".equals(userPasswordDemoLoginData.password)
                && "user".equals(userPasswordDemoLoginData.user) && userPasswordDemoLoginData.isDemo == true;

        if (isValid) {
            // Report succesful login
            adminListeners.forEach(Layer1ApiAdminListener::onLoginSuccessful);

            // Generate some events each second
            while (!Thread.interrupted()) {
                
                // Generate some data changes
                simulate();

                // Waiting a bit before generating more data
                try {
                    Thread.sleep(1000);
                } catch (@SuppressWarnings("unused") InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        } else {
            // Report failed login
            adminListeners.forEach(l -> l.onLoginFailed(LoginFailedReason.WRONG_CREDENTIALS,
                    "This provider only acepts following credentials:\n"
                            + "username: user\n"
                            + "password: pass\n"
                            + "is demo: checked"));
        }
    }

    protected void simulate() {
        // Generating some data for each of the instruments
        synchronized (instruments) {
            instruments.values().forEach(Instrument::generateData);
        }
    }

    @Override
    public String getSource() {
        // String identifying where data came from.
        // For example you can use that later in your indicator.
        return "realtime demo";
    }

    @Override
    public void close() {
        // Stop events generation
        connectionThread.interrupt();
    }

}

File: cpp/FullTextDataRealtimeProvider/DemoExternalRealtimeProvider.cpp
// Demo to use with FullTextDataRealtimeProvider that
// generates random data for instruments that you subscribe to.
// You can compile it with visual studio by launching
// "Developer Command Prompt" (e.g. for VS 2017 it's
// "Developer Command Prompt for VS 2017" and running "cl DemoExternalRealtimeProvider.cpp"
// in the folder with this cpp file.
// This code was tested under Windows with VS 2017, but you should be able to use similar
// (or exactly the same) approach with other platforms/compilers

#include <iostream>
#include <thread>
#include <chrono>
#include <map>
#include <mutex>
#include <string>
#include <cstdio>
#include <cstdlib>

using namespace std;

const int DEPTH_LEVELS_COUNT = 10;

// We don't really need to report time
// (bookmap will ignore it in realtimemode),
// but just for consistency
long long getTime() {
	long long ms = chrono::duration_cast<chrono::nanoseconds>(
		chrono::system_clock::now().time_since_epoch()).count();
	return ms;
}

const char * toString(bool b) {
	return b ? "true" : "false";
}

// Primitive way to generate JSON. Will break if strings contain some
// characters, like " or \n. Used for simplicity.
void onDepth(string alias, bool isBid, int price, int size) {
	printf("Depth {\"alias\":\"%s\",\"isBid\":%s,\"price\":%d,\"size\":%d,\"time\":%lld}\n",
		alias.c_str(), toString(isBid), price, size, getTime());
}

void onTrade(string alias, bool isBid, int price, int size) {
	printf("Trade {\"alias\":\"%s\",\"price\":%d,\"size\":%d,\"tradeInfo\":{\
\"isOtc\":false,\"isBidAggressor\":%s,\"isExecutionStart\":true,\"isExecutionEnd\":true}\
,\"time\":%lld}\n",
		alias.c_str(), price, size, toString(isBid), getTime());
}

void onInstrumentAdded(string alias, string symbol, string exchange, string type,
	double pips, double multiplier, string fullName, bool isFullDepth,
	double sizeMultiplier) {
	printf("InstrumentAdded {\"alias\":\"%s\",\"instrumentInfo\":\
{\"pips\":%lf,\"multiplier\":%lf,\"fullName\":\"%s\",\"isFullDepth\":%s,\
\"sizeMultiplier\":%lf,\"symbol\":\"%s\",\"exchange\":\"%s\",\"type\":\"%s\"},\
\"time\":%lld}\n",
		alias.c_str(), pips, multiplier, fullName.c_str(), toString(isFullDepth),
		sizeMultiplier, symbol.c_str(), exchange.c_str(), type.c_str(), getTime());
	// Flushing because otherwise user will have to wait while it's in buffer
	fflush(stdout);
}

void onInstrumentAlreadySubscribed(string symbol, string exchange, string type) {
	printf("InstrumentAlreadySubscribed {\"symbol\":\"%s\",\"exchange\":\"%s\",\
\"type\":\"%s\",\"time\":%lld}\n",
	symbol.c_str(), exchange.c_str(), type.c_str(), getTime());
	// Flushing because otherwise user will have to wait while it's in buffer
	fflush(stdout);
}

void onLoginSuccessful() {
	printf("LoginSuccessful {\"time\":%lld}\n", getTime());
	// Flushing because otherwise user will have to wait while it's in buffer
	fflush(stdout);
}

void onLoginFailed(string reason, string message) {
	printf("LoginFailed {\"reason\":\"%s\",\"message\":\"%s\",\"time\":%lld}\n",
		reason.c_str(), message.c_str(), getTime());
	// Flushing because otherwise user will have to wait while it's in buffer
	fflush(stdout);
}

struct Instrument {

	string alias;
	double pips;

	int basePrice;

	Instrument(string alias, double pips) {
		this->alias = alias;
		this->pips = pips;

		// Pick random price that will be used to generate the data
		// This is an integer representation of a price (before multiplying
		// by pips)
		this->basePrice = (int)(rand() % 10000 + 1000);
	}

	void generateData() {

		// Determining best bid/ask
		int bestBid = getBestBid();
		int bestAsk = getBestAsk();

		// Populating 10 levels to each side of best bid/best ask with
		// random data
		for (int i = 0; i < DEPTH_LEVELS_COUNT; ++i) {
			int levelsOffset = i;
			onDepth(alias, true, bestBid - levelsOffset, getRandomSize());
			onDepth(alias, false, bestAsk + levelsOffset, getRandomSize());
		}

		// Trade on best bid, ask agressor
		onTrade(alias, false, bestBid, 1);
		// Trade on best ask, bid agressor
		onTrade(alias, true, bestAsk, 1);

		// With 10% chance change BBO
		if (rand() % 100 < 10) {
			// 50% chance to move up, 50% to move down
			if (rand() % 100 > 50) {
				// Moving up - erasing best ask, erasing last reported bid
				// level (emulating exchange only reporting few levels)
				++basePrice;
				onDepth(alias, false, bestAsk, 0);
				onDepth(alias, true, bestBid - (DEPTH_LEVELS_COUNT - 1), 0);
				// Could also populate new best bid and add last best ask,
				// but this can be omitted - those will be populated during
				// next simulation step
			}
			else {
				// Moving down - erasing best bid, erasing last reported ask
				// level (emulating exchange only reporting few levels)
				--basePrice;
				onDepth(alias, true, bestBid, 0);
				onDepth(alias, false, bestAsk + (DEPTH_LEVELS_COUNT - 1), 0);
				// Could also populate new best ask and add last best bid,
				// but this can be omitted - those will be populated during
				// next simulation step
			}
		}
	}

	int getBestAsk() {
		return basePrice;
	}

	int getBestBid() {
		return getBestAsk() - 1;
	}

	int getRandomSize() {
		return (int)(1 + rand() % 10);
	}

};

bool closing = false;

mutex instrumentsMutex;
map<string, Instrument> instruments;

void login(string user, string password, bool demo) {
	// With real connection provider would attempt establishing connection here.
	bool isValid = "pass" == password && "user" == user && demo == true;

	if (isValid) {
		// Report succesful login
		onLoginSuccessful();
	}
	else {
		// Report failed login
		// Since we don't have proper json serializer in this example,
		// we have to escape newlines here
		onLoginFailed("WRONG_CREDENTIALS",
			"This provider only acepts following credentials:\\n\
username: user\\npassword: pass\\nis demo: checked");
	}
}

string createAlias(string symbol, string exchange, string type) {
	return symbol + "/" + exchange + "/" + type;
}

void subscribe(string symbol, string exchange, string type) {
	std::lock_guard<std::mutex> lock(instrumentsMutex);

	string alias = createAlias(symbol, exchange, type);
	if (instruments.find(alias) != instruments.end()) {
		onInstrumentAlreadySubscribed(symbol, exchange, type);
	}
	else {
		// We are performing subscription synchronously for simplicity,
		// but if subscription process takes long it's better to do it
		// asynchronously

		// Randomly determining pips. In reality it will be received
		// from external source
		double pips = rand() % 100 > 50 ? 0.5 : 0.25;

		instruments.emplace(alias, Instrument(alias, pips));

		onInstrumentAdded(alias, symbol, exchange, type, pips, 1, "Full name:" + alias, false, 1);
	}
}

void unsubscribe(string alias) {
	std::lock_guard<std::mutex> lock(instrumentsMutex);
	instruments.erase(alias);
}

void simulateStep() {
	std::lock_guard<std::mutex> lock(instrumentsMutex);
	for (auto entry : instruments) {
		entry.second.generateData();
	}
	// Instead of flushing every single update we flush at the end
	// This is a bit more performance-friendly approach
	// But this probably doesn't make a practical difference
	// as 200K+ events per second can be pushed both ways,
	// bottlenecking the actual processing.
	fflush(stdout);
}

void simulate() {
	while (!closing) {
		// Generate some data changes
		simulateStep();
		// Waiting a bit before generating more data.
		// Interruptable sleep would be better.
		// Downside of this approach is that bookmap will have to wait up to 1s to
		// shut down this thread.
		this_thread::sleep_for(chrono::milliseconds(1000));
	}
}

int main() {

	thread simulationThread(simulate);

	bool readSuccess;
	string command;
	do {
		readSuccess = !getline(cin, command).eof();
		if (!readSuccess) {
			// Do nothing. This is end of file.
			// (bookmap died without closing the module). 
			// If we ignore EOF subprocess might stay e.g.
			// if bookmap was killed via task manager.
		} else if (command == "login") {
			
			string user, password, demoString;
			getline(cin, user);
			getline(cin, password);
			getline(cin, demoString);
			bool demo = demoString == "true";

			login(user, password, demo);

		} else if (command == "subscribe") {

			string symbol, exchange, type;
			getline(cin, symbol);
			getline(cin, exchange);
			getline(cin, type);

			subscribe(symbol, exchange, type);

		} else if (command == "unsubscribe") {
			string alias;
			getline(cin, alias);
			unsubscribe(alias);
		}
	} while (command != "close" && readSuccess);

	closing = true;
	simulationThread.join();

	return 0;
}
